project_liberation/requirements.txt

```diff
@@ -0,0 +1,9 @@
+black
```

Comment:
Why not [pipenv](https://docs.pipenv.org)?

Reply:
Have you had a chance to use it already?

Reply:
I used it for a brief time at the beginning of one of our projects but back then it had too many issues (e.g. it was dumping system-specific debug info into the lock file). But these issues have been resolved for a long time now and I've been trying to encourage people to try it several times. Everyone complains about the requirements file and people do the same mistakes over and over again but for some weird reason no one wants to try using a better tool when given a chance to do so in a new project.

Reply:
OK, then I would be for using it too. Just wanted to know if you have used it already so we don't run into some problems later.

Reply:
I have replaced requirements with pipfiles to try it, it has some problems with installing `black` but it is known problem https://github.com/microsoft/vscode-python/issues/5171. After that, it is working on fresh virtualenv, so I have added it to this branch in separate commit.

Reply:
I think this thread is more relevant: https://github.com/python/black/issues/517. The VSCode solution seems to be to change the way you run pipenv (i.e. to run it twice and add `--pre` flag when running it just for black if I understand correctly?). Pinning to a specific version as suggested in the thread above seems like a better workaround to me (i.e. something like `black = "==18.9b0"`).

Anyway, this is not specific to black - any dependency. without a stable release will have this issue. But it's weird that they don't have one at this stage.

Reply:
I see that you actually used `allow_prereleases`. I don't think it's a good idea in general. At most as a workaround but then it should have a comment pointing out that it's only temporary. Anyway, I'd personally just pin the version.

Reply:
I get some weird results with this tool. Once suddenly, today it started working without `allow_prereleases` on fresh venv. Then, repeating the process, I get back to the original error.

At some point when I have set specific version for `black` I get to the error 
```
pipenv.patched.notpip._internal.exceptions.UnsupportedPythonVersion: black requires Python '>=3.6' but the running Python is 2.7.12
``` 

when I have set it to `==19.3b0` even when I am on virtualenv with Python 3.6. Setting back to `black = "*"` or `==18.9b0` allowed me to install. 

For now it works on fresh venv but I am a little bit sceptical if we won't run into some problems with it.

Reply:
Maybe it's actually a problem with black? Or at least with the newest release?

Anyway, I don't really know how reliable pipenv is but the best way to find out is to start using it. A small project like this is the perfect opportunity to do so. If it turns out it has serious drawbacks, we won't be using it for anything else. But if it's good, we'll finally be able to stop doing the tedious and error-prone dependency management manually so it's worth trying.

By the way, I was under impression that pipenv was becoming a standard tool in the Python community but seems like lots of people don't like it for various reasons. The most common arguments against it are related to the fact that it does not replace pip and does not work well for libraries - which is not a problem for us since we have an application here. You could try one of the alternatives:
- [poetry](https://poetry.eustace.io/)
- [hatch](https://github.com/ofek/hatch)
- [flit](https://github.com/takluyver/flit)

Reply:
I am not saying I don't like it, just that I was very excited at the beginning, and now I am little less. But lets keep it. Maybe its problem with black itself as you have said.